src/lib/ndpi_main.c

```diff
@@ -11423,3 +11423,30 @@ char *ndpi_dump_config(struct ndpi_detection_module_struct *ndpi_str,
   }
   return NULL;
 }
+
+void* ndpi_memmem(const void* haystack, size_t haystack_len, const void* needle, size_t needle_len)
+{
+  if (!haystack || !needle) {
+    return NULL;
+  }
+
+  if (needle_len > haystack_len) {
+    return NULL;
+  }
+
+  if ((size_t)(const u_int8_t*)haystack+haystack_len < haystack_len) {
+    return NULL;
+  }
+
+  if (needle_len == 1) {
+    return memchr(haystack, *(int*)needle, haystack_len);
+  }
+
+  const u_int8_t* h = NULL;
+  for (h = haystack; h <= (const u_int8_t*)haystack+haystack_len-needle_len; ++h) {
+    if (!memcmp(h, needle, needle_len))
```

Comment:
Something like (not tested!)?
```
if(*h == needle[0] && !memcmp(h, needle, needle_len))
```
To avoid too many memcmp

Reply:
I guess it could be added, but I didn't see any noticeable change. Maybe my benchmark is wrong, or maybe the difference is less than 1 ns.

[memmem_bench.zip](https://github.com/ntop/nDPI/files/14931089/memmem_bench.zip)

![screenshot1](https://github.com/ntop/nDPI/assets/105977161/943f474d-634a-44d3-9a0e-2fe40668ac84)
![screenshot2](https://github.com/ntop/nDPI/assets/105977161/7fa841d3-200e-4f92-ab02-2cdbf0beafc1)